### PR TITLE
Reject writer.ready promise on errors, instead of fulfilling

### PR DESCRIPTION
**Pull request by domenic**
In the process of testing this, ported several tests to web-platform-tests format, and introduced the recording stream helpers that I am using extensively in #512.

---

Apologies if we discussed this before and came to a good reason for this behavior; I couldn't find anything by searching. Anyway, this seems much more intuitive to me... the previous behavior caused some of my pipe code to fail since I wrote the code expecting that ready would only be fulfilled when it was safe to write.

### CODE REVIEW

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  self.importScripts('/resources/testharness.js');
+  self.importScripts('../resources/test-utils.js');
+  self.importScripts('../resources/recording-streams.js');
+}
+
+const error1 = new Error('error1');
+error1.name = 'error1';
+
+promise_test(t => {
+
```
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: Unnecessary blank line?

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/service-workers/service-worker/resources/test-helpers.sub.js"></script>
+<script src="../resources/test-utils.js"></script>
```
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html]: Doesn't appear to be used? Also in importScripts bad-underlying-sinks.js below.

File: reference-implementation/to-upstream-wpts/resources/test-utils.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+'use strict';
```
Comment (ricea) [inline reference-implementation/to-upstream-wpts/resources/test-utils.js]: This file doesn't seem to be used in this commit, so I would prefer for it to be landed in a separate commit.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  const writer = ws.getWriter();
+
+  return promise_rejects(t, error1, writer.close(), 'close() promise must reject with the thrown error')
+  .then(() => promise_rejects(t, error1, writer.ready, 'ready promise must reject with the thrown error'))
+  .then(() => {
+    assert_array_equals(ws.events, ['close']);
+  });
+
+}, 'Underlying sink close: throwing method');
```
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: I prefer "should" phrasing in test descriptions, eg. 'close() and ready promises should reject when sink close() throws'.

Also I would prefer it if the test descriptions didn't contain prefixes than can be inferred from the context.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        return delay(10);
+      }
+
+      return Promise.reject(error1);
+    }
+  });
+
+  const writer = ws.getWriter();
+
+
```
Comment (tyoshino) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: single blank line

File: reference-implementation/test/writable-stream.js [hunk 6]
```diff
@@ -473,154 +473,6 @@ test('If close is called on a WritableStream in waiting state, ready will be ful
   }, 0);
 });
 
-test('If sink rejects on a WritableStream in writable state, ready will return a fulfilled promise', t => {
```
Comment (tyoshino) [inline reference-implementation/test/writable-stream.js]: you've dropped this since the "returning a rejected promise (second write)" test would cover most of this?

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js (hunk 1 above)
Comment (domenic) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: Personally I write the tests with blank lines after the opening brace and before the last line/description. Otherwise it's too hard to distinguish the test description from the test code. *shrug*

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js (hunk 4 above)
Comment (domenic) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: Hmm. For certain types of tests I prefer the description to describe the situation being tested, and the assertions themselves to contain the "must"s. Let me tweak this though since I guess most tests are moving in that direction.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html (hunk 2 above)
Comment (domenic) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html]: It's used for delay()

File: reference-implementation/to-upstream-wpts/resources/recording-streams.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+"use strict";
+
+self.recordingReadableStream = (extras = {}, strategy) => {
+  let controllerToCopyOver;
+  const stream = new ReadableStream({
+    start(controller) {
+      controllerToCopyOver = controller;
+
+      if (extras.start) {
```
Comment (domenic) [inline reference-implementation/to-upstream-wpts/resources/recording-streams.js]: Oh gross. Definitely not in this repo; I think we have a lint rule disallowing such elision :)

File: reference-implementation/test/writable-stream.js (hunk 6 above)
Comment (domenic) [inline reference-implementation/test/writable-stream.js]: Hmm, I guess it might be worth porting this test more directly as well. Will do.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js (hunk 1 above)
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: I am biased towards Chromium style. That may or may not correspond with what is appropriate for web-platform-tests.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js (hunk 4 above)
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.js]: I found when I was porting tests that it was easy to see what a test was for when it was phrased in terms of requirements. I think this is important when looking for gaps in testing coverage.

File: reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html (hunk 2 above)
Comment (ricea) [inline reference-implementation/to-upstream-wpts/writable-streams/bad-underlying-sinks.https.html]: Sorry. I missed that.

File: reference-implementation/to-upstream-wpts/resources/test-utils.js (hunk 3 above)
Comment (ricea) [inline reference-implementation/to-upstream-wpts/resources/test-utils.js]: Sorry, I missed the use of "delay". Please disregard this comment.

File: reference-implementation/to-upstream-wpts/resources/recording-streams.js (hunk 7 above)
Comment (ricea) [inline reference-implementation/to-upstream-wpts/resources/recording-streams.js]: I will keep that in mind, thank you.

**Comment by tyoshino** [comment]
e69528c lgtm
